substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/thread/JavaThreads.java

```diff
@@ -430,6 +441,178 @@ static void setCurrentThread(Thread carrier, Thread thread) {
         currentVThreadId.set(getThreadId(thread));
     }
 
+    public static class JMXMonitoring {
+        private static final ThreadMXBean THREAD_MX_BEAN = ManagementFactory.getThreadMXBean();
+        private static final int NO_VALUE = -1;
+
+        @Uninterruptible(reason = CALLED_FROM_UNINTERRUPTIBLE_CODE, mayBeInlined = true)
+        public static void handleContentionMonitoring(Target_java_lang_Thread targetThread, int threadStatus) {
+            boolean contentionEnabled = THREAD_MX_BEAN.isThreadContentionMonitoringEnabled();
+            if (contentionEnabled) {
+                handleContentionWhenEnabled(targetThread, threadStatus);
+            } else {
+                resetContentionIfStarted(targetThread);
+            }
+        }
+
+        @Uninterruptible(reason = CALLED_FROM_UNINTERRUPTIBLE_CODE, mayBeInlined = true)
+        private static void resetContentionIfStarted(Target_java_lang_Thread targetThread) {
+            if (targetThread.lastStartedBlocked != NO_VALUE || targetThread.lastStartedWaiting != NO_VALUE) {
+                targetThread.lastStartedBlocked = targetThread.lastStartedWaiting = NO_VALUE;
+            }
+        }
+
+        @Uninterruptible(reason = CALLED_FROM_UNINTERRUPTIBLE_CODE, mayBeInlined = true)
+        private static void handleContentionWhenEnabled(Target_java_lang_Thread targetThread, int threadStatus) {
+            switch (threadStatus) {
+                case ThreadStatus.BLOCKED_ON_MONITOR_ENTER:
+                    setThreadBlockedToEnterLock(targetThread);
+                    break;
+                case ThreadStatus.IN_OBJECT_WAIT:
+                case ThreadStatus.IN_OBJECT_WAIT_TIMED:
+                case ThreadStatus.SLEEPING:
+                case ThreadStatus.PARKED:
+                case ThreadStatus.PARKED_TIMED:
+                    setThreadStartsWaiting(targetThread);
+                    break;
+                case ThreadStatus.RUNNABLE:
+                    calculateContentionOnThreadRunning(targetThread);
+                    break;
+            }
+        }
+
+        @Uninterruptible(reason = CALLED_FROM_UNINTERRUPTIBLE_CODE, mayBeInlined = true)
+        private static void setThreadBlockedToEnterLock(Target_java_lang_Thread thread) {
+            if (thread.lastStartedBlocked == NO_VALUE) {
+                thread.lastStartedBlocked = System.currentTimeMillis();
+                thread.blockedCount++;
```

Comment:
Are we guaranteed to always be dealing with the current thread? If so, maybe it's worth adding an assertion here. 

Otherwise, is it safe to be un-atomically modifying `thread.lastStartedBlocked`, `thread.blockedCount`, `thread.waitedCount`, etc ? Or do we require synchronization here as well?

Reply:
For the thread states that I'm interested in, it's modified by the current thread only. 
However, there are use-cases such as when a new thread is initialized, then its state can be set by its parent just before the new thread is assigned as current. In this case, the thread state will be set to `NEW`, and I'm not updating anything for this state.